doc/topics/async-system.md

```diff
@@ -0,0 +1,209 @@
+
+Imagine that we're using a third-party library that does asynchronous work with a simple callback-based API. The library's function called `computeSomethingSlowly` does some work in the background and then calls the callback that we pass to it:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp compute-something-slowly
+
+We'd like to be able to use the `AsyncSystem` to do further work after this completes. Here's how that can be done:
+
+\snippet{trimleft,strip} ExamplesAsyncSystem.cpp compute-something-slowly-async-system
+
+The trick is to call [createPromise](@ref CesiumAsync::AsyncSystem::createPromise) to create a `Promise<T>`, and then capture it in the callback lambda given to the library's `computeSomethingSlowly`. When the callback is invoked, we call [resolve](@ref CesiumAsync::Promise::resolve) on the `Promise<T>`, which resolves the assocated `Future<T>` that was previously obtained with [getFuture](@ref CesiumAsync::Promise::getFuture). We can then chain continuations off that `Future<T>` in the normal way. Instead of calling `resolve` on the `Promise<T>`, we can also call [reject](@ref CesiumAsync::Promise::reject).
+
+One problem with this `createPromise` approach is that it does not behave as well as it should in the face of exceptions. Imagine that we're wrapping all of this functionality up in our own function, like this:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp compute-something-slowly-wrapper
+
+Now what happens if that third-party library function, `computeSomethingSlowly`, throws an exception? Our `myComputeSomethingSlowlyWrapper` doesn't catch it, so the exception will be propagated out of our function and the caller will have to deal with it. This is poor form in a function that returns a `Future<T>`, because it forces callers to handle two different error mechanisms. The function might throw an exception, or the function might reject the `Future<SlowValue>` that it returns. Handling both would be a hassle, and it's unfair to ask our users to do so.
+
+We could change `myComputeSomethingSlowlyWrapper` to catch the possible exception and turn it into a call to `reject` on the `Promise<T>`, and that would work fine. But there's an easier way. We can use the [createFuture](@ref CesiumAsync::AsyncSystem::createFuture) method on `AsyncSystem`:
+
+\snippet{trimleft} ExamplesAsyncSystem.cpp compute-something-slowly-wrapper-handle-exception
+
+The `createFuture` method takes a function as its only parameter, and it _immediately_ invokes that function (that's why it's safe to capture everything by reference `&` in this case). The function receives a parameter of type `Promise<T>`, which is identical to the one that would be created by `createPromise`. Just like before, we call the library function that we're wrapping and resolve the `Promise<T>` when its callback is invoked.
```

Comment:
If I understand correctly, the function called by `createFuture()` need to explicitly schedule work with e.g. `runInWorkerThread` if it could block. If that's true, then I think it needs to be mentioned here.

Reply:
Well... yes and no. `createFuture` calls its callback immediately in the same thread (I updated the text to say this), so if you plan to do heavy CPU work, you certainly shouldn't do it there. In that scenario, you should call `runInWorkerThread` _instead_, no need for `createFuture` at all.

However, if you're planning to block (rather than just do heavy CPU work)... well, maybe don't do that? Blocking is bad, even if it's a worker thread you're blocking. It's usually possible to avoid. But if you truly must block, then that's another case for using `runInWorkerThread` and there's no need to use `createFuture`.

The real purpose of `createFuture`, though, is to interface with async APIs that use a different mechanism for notifying of completion, usually a callback. That's the example shown here. In that scenario, the async API probably already avoids doing heavy work in the calling thread, so there's no reason to start a worker thread for it.

Reply:
> Well... yes and no. `createFuture` calls its callback immediately in the same thread (I updated the text to say this), so if you plan to do heavy CPU work, you certainly shouldn't do it there. In that scenario, you should call `runInWorkerThread` _instead_, no need for `createFuture` at all.
> 
> However, if you're planning to block (rather than just do heavy CPU work)... well, maybe don't do that? Blocking is bad, even if it's a worker thread you're blocking. It's usually possible to avoid. But if you truly must block, then that's another case for using `runInWorkerThread` and there's no need to use `createFuture`.
> 

Too bad I didn't have this guide when I wrote UrlAccessor :sunglasses:  I understand now that you can just return the result of `runInWorkerThread` without wrapping that call in `createFuture`.